File: frontend/src/backend/itemsService.js
import {config, appFetch} from './appFetch';

export const createItem = (formData, onSuccess, onErrors) =>
    appFetch(`/items/createItem`,
        config('POST', formData), onSuccess, onErrors);

export const checkInventory = ({keywords, page}, onSuccess, onErrors) => {

    let path = `/items/checkInventory?page=${page}`;

    if(keywords) {
        path +=`&keywords=${encodeURIComponent(keywords)}`;
    }

    appFetch(path, config('GET'), onSuccess, onErrors);
}

export const addItemBoxToWarehouse = (itemId, numItems, warehouseName, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}/addItemBoxToWarehouse`,
        config(`POST`, {numItems, warehouseName}), onSuccess, onErrors);

export const findItemById = (itemId, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}`,
        config('GET'), onSuccess, onErrors);

export const countNumBoxesOfItemId = (itemId, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}/numBoxes`,
        config('GET'), onSuccess, onErrors);

export const findAllBoxesOfItemId = (itemId, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}/boxes`,
        config('GET'), onSuccess, onErrors);

export const deleteItem = (itemId, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}/deleteItem`,
        config('POST'), onSuccess, onErrors);

export const modifyItem = (itemId, formData, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemId}/modifyItem`,
        config('PUT', formData), onSuccess, onErrors);

export const modifyItemBox = (itemBoxId, numItems, warehouseName, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemBoxId}/modifyItemBox`,
        config('PUT', {numItems, warehouseName}), onSuccess, onErrors);

export const addItemsToBox = (itemBoxId, numItemsToAdd, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemBoxId}/addItemsToBox?numItemsToAdd=${numItemsToAdd}`,
        config('PUT'), onSuccess, onErrors);

export const removeItemsFromBox = (itemBoxId, numItemsToRemove, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemBoxId}/removeItemsFromBox?numItemsToRemove=${numItemsToRemove}`,
        config('PUT'), onSuccess, onErrors);

export const deleteItemBox = (itemBoxId, onSuccess, onErrors) =>
    appFetch(`/items/checkInventory/${itemBoxId}/deleteItemBox`,
        config('POST'), onSuccess, onErrors);

export const createSupplier = (supplierName, onSuccess, onErrors) =>
    appFetch(`/items/suppliers/create`,
        config('POST', {supplierName}), onSuccess, onErrors);

export const findAllSuppliers = (onSuccess, onErrors) =>
    appFetch(`/items/suppliers/findSuppliers`,
        config('GET'), onSuccess, onErrors);

export const findItemsFromSupplier = ({supplierId, page}, onSuccess, onErrors) =>
    appFetch(`/items/suppliers/${supplierId}/findItemsFromSupplier?page=${page}`,
        config('GET'), onSuccess, onErrors);

export const findSupplierById = (supplierId, onSuccess, onErrors) =>
    appFetch(`/items/suppliers/${supplierId}`,
        config('GET'), onSuccess, onErrors);

export const findAllItems = (onSuccess, onErrors) =>
    appFetch(`/items`, config('GET'), onSuccess, onErrors);

File: backend/src/main/java/backend/model/services/ItemsService.java
package backend.model.services;

import backend.model.entities.Item;
import backend.model.entities.ItemBox;
import backend.model.entities.Supplier;
import backend.model.exceptions.InstanceNotFoundException;
import backend.model.exceptions.InvalidNumberOfItemsToAddToBox;
import backend.model.exceptions.InvalidNumberOfItemsToRemoveFromBox;
import backend.model.exceptions.PermissionException;

import java.util.List;

public interface ItemsService {

    Item createItem(Long userId, String itemName, String referenceCode, String barCode, String manufacturerRef,
                    String supplier, byte[] imgFile) throws PermissionException, InstanceNotFoundException;

    Long addItemBoxToWarehouse(Long userId, Long itemId, Long numItems, String warehouseName)
            throws PermissionException, InstanceNotFoundException;

    Block<Item> checkInventory(Long userId, String keywords, int page, int size) throws PermissionException, InstanceNotFoundException;

    Item findItemById(Long userId, Long itemId) throws PermissionException, InstanceNotFoundException;

    Long countNumBoxesOfItemId(Long userId, Long itemId) throws PermissionException, InstanceNotFoundException;

    List<ItemBox> findAllBoxesOfItemId(Long userId, Long itemId) throws PermissionException, InstanceNotFoundException;

    Boolean deleteItem(Long userId, Long itemId) throws PermissionException, InstanceNotFoundException;

    Long modifyItem(Long userId, Long itemId, String itemName, String referenceCode, String barCode,
                    String manufacturerRef, String supplier, byte[] imgFile)
            throws PermissionException, InstanceNotFoundException;

    Long modifyItemBox(Long userId, Long itemBoxId, Long numItems, String warehouseName)
            throws PermissionException, InstanceNotFoundException;

    Long addItemsToBox(Long userId, Long itemBoxId, Long numItemsToAdd)
            throws PermissionException, InstanceNotFoundException, InvalidNumberOfItemsToAddToBox;

    Long removeItemsFromBox(Long userId, Long itemBoxId, Long numItemsToRemove)
            throws PermissionException, InstanceNotFoundException, InvalidNumberOfItemsToRemoveFromBox;

    Boolean deleteItemBox(Long userId, Long itemBoxId) throws PermissionException, InstanceNotFoundException;

    Supplier createSupplier(String supplierName);

    List<Supplier> findAllSuppliers();

    Block<Item> findItemsFromSupplier(Long supplierId, int page, int size) throws InstanceNotFoundException;

    Supplier findSupplierById(Long supplierId) throws InstanceNotFoundException;

    List<Item> findAllItems();

}
